### DISCUSSION TITLE
Using native-compile for borg-compile-function doesn't seem to work as advertised

### DISCUSSION THREAD

**Discussion by league** [emacscollective/borg#125] [Support] [answered]
Hello! I just bootstrapped from the emacs.g seed, and updated borg drone to the latest, cd1fbd60a2ff62bbfca124ea1e602b14c4aa58e9.  When I run `make`, it will generate the `.elc` files in all the `lib/DRONE` locations, and when they are first loaded emacs 28.1 will native-compile them. But I'd like to do the native-compile ahead of time.

I tried this setting in `etc/borg/config.el`, as suggested in the borg manual:
```el
(when (native-comp-available-p)
  (setq borg-compile-function #'native-compile))
```

With that, `make` says it's (re-)building everything (and maybe it is?), but afterward, there are no more `.elc` files in the `lib/DRONE` locations. I **think** that when the `.elc` is missing, Emacs doesn't bother to load the native `.eln` either!

When I restart emacs, I try these to verify – loading one function provided with emacs and one from a drone:
```elisp
(describe-function 'font-lock-refresh-defaults)
;; ⇒ font-lock-refresh-defaults is a native compiled Lisp function in ‘font-lock.el’
(describe-function 'epkg-describe-package)
;; ⇒ epkg-describe-package is an autoloaded interactive Lisp closure in ‘epkg-desc.el’
```
The epkg function is not reported as compiled at all, even byte-compiled.

One hacky way I've found to resolve this is to set `borg-compile-function` so it explicitly does **both:**

```elisp
(when (native-comp-available-p)
  (setq borg-compile-function
	(defun cl/my-compile (filename)
	  (and (byte-compile-file filename)
	       (native-compile filename)))))
```

Now after a `make` the `.elc` files stick around, and the native files seem to be ready ahead of time and load properly:

```elisp
(describe-function 'epkg-describe-package)
;; ⇒ epkg-describe-package is an interactive native compiled Lisp function in ‘epkg-desc.el’.
```

So maybe this is a documentation bug about that `config.el` setting, or borg itself needs to ensure the `.elc` is also created when it's `native-compile`, or maybe there's something strange about my setup (or my interpretation of it)? Thanks!

**Comment by Eason0210** [comment]
@tarsius 

I can reproduce this issue. 

The workaround also works for me
```emacs-lisp
(when (native-comp-available-p)
  (setq borg-compile-function
	(defun cl/my-compile (filename)
	  (and (byte-compile-file filename)
	       (native-compile filename)))))
```
And  the issue on building packages with `compat.el` also fixed.
https://github.com/magit/magit/issues/4665

**Comment by tarsius** [comment]
> Using native-compile for borg-compile-function doesn't seem to work as advertised

I don't think I *advertised* it. I always said you use it at your own risk. I only added support because people insisted it was a good idea and it looked like it more or less would just worked (but now I think it never really worked). But I guess now I am on the hook and have to implement it properly.

See the `native` branch. Still needs more testing, tweaking, foot-gun removal and documentation; but otherwise is ready.

- Use `make clean build-native` to wait a long time,
- or use `make clean all; emacs; make native-compile` to start using emacs as soon as the byte-code is ready.
- Use `(setq borg-compile-function #'borg-byte+native-compile)` to make the `borg-compile` command slower,
- or don't do that and let Emacs continue to perform the native compilation asynchronously.

**Reply by tarsius** [reply]
Still not done, but seems usable. This is in the `work` branch now.

**Reply by Eason0210** [reply]
👍  Thanks for your great work on it. 
I tried the `work` branch with `make native`, It works well as expected.
And I aslo tried to set the `/.emacs.d/etc/borg/config.el` as below:
```
(when (native-comp-available-p)
  (setq borg-compile-function #'native-compile))
```
Then <kbd>M-x borg-assimilate</kbd> `<package-name>`, It also works well.

Tested on macOS with Emacs 29.0.50

**Reply by tarsius** [reply, marked as answer]
I've improved the implementation a while ago and believe it now works. (But again cannot test because I cannot use native compilation due to complications that have nothing to do with Borg.)
